### PR TITLE
Add leisure

### LINKED ISSUE #8186: [BUG] Fisherman endlessly cycles without fishing
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Are you using the latest MineColonies Verison?

- [X] I am running the latest **alpha** version of MineColonies for my Minecraft version.

### Did you check on the Wiki? or ask on Discord?

- [X] I checked the MineColonies Wiki and made sure my issue is not covered there. Or I was sent from discord to open an issue here.

### Minecraft Version

1.16

### MineColonies Version

1.0.540-ALPHA

### Structurize Version

1.0.371-ALPHA

### Related Mods and their Versions

Forge 36.2.29

### Current Behavior

The fisherman walks to their hut block, walks to the center of a pool, walks back to their hut block, repeating forever.  They don't actually do any fishing.

### Expected Behavior

They actually do some fishing.

### Reproduction Steps

Give the fisherman precisely a 7x7 pool of water, with no other pools nearby, and surrounded on all but one side with fences.

The AI happens to pick a water block in the center of one of the fence sides as the "pond" location.  (The "standing location" that it picked also is a water block; not sure if that's random or if it just picked something that wasn't the fence.)

https://github.com/ldtteam/minecolonies/blob/8fca9b1f8b0b1fc321e65b720f172a4178c341d9/src/main/java/com/minecolonies/coremod/entity/ai/citizen/fisherman/EntityAIWorkFisherman.java#L313-L316
At this point, it detects that the chosen block is water, then it resets to retry the whole state machine again -- except that it hasn't changed the pond or any other conditions, so it will just go right back here again indefinitely.

Possibly the "real bug" is that the pond search reported two water blocks and not a good standing location, but this reset doesn't seem to be resetting properly.  Or perhaps I'm missing something, but it's not right as is.

### Logs

n/a

### Anything else?

Might be nice to have it look for a "work" tag as a preferred place to stand while fishing (though perhaps still searching for others).

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### LINKED ISSUE #8189: [BUG] Crashed loading an existing world
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Are you using the latest MineColonies Verison?

- [X] I am running the latest **alpha** version of MineColonies for my Minecraft version.

### Did you check on the Wiki? or ask on Discord?

- [X] I checked the MineColonies Wiki and made sure my issue is not covered there. Or I was sent from discord to open an issue here.

### Minecraft Version

1.16

### MineColonies Version

1.0.540-ALPHA

### Structurize Version

1.0.371-ALPHA

### Related Mods and their Versions

1. Forge Version: 36.2.33

### Current Behavior

On Loading, it suddenly crashes.

### Expected Behavior

It should have loaded successfully.

### Reproduction Steps

1. Start Minecraft 1.16.5
2. Load an existing world

### Logs

https://gist.github.com/CaptainAhmed18/483ad4491181b7aaadb0c99ddb09088c

### Anything else?

[crash-2022-03-27_11.36.00-server.txt](https://github.com/ldtteam/minecolonies/files/8357218/crash-2022-03-27_11.36.00-server.txt)

[latest.log](https://github.com/ldtteam/minecolonies/files/8357217/latest.log)

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #8186
Closes #8189
Closes #

# Changes proposed in this pull request:
- Adds leisure to the random wander job.
- Citizens go to leisure places, hang out, sit around, etc.
- Adjust mourning handling
- Citizens go to graveyard and have a chance to reduce mourning time (also walk to the grave of the diseased)
- Fix fishing
- Adjust position of the child overhead icon.
- Remove unnecessary shuffle methods.

Still waiting for some more schematic updated

Review please

### CODE REVIEW

**Review by marchermans — DISMISSED** [review]

File: src/main/java/com/minecolonies/coremod/entity/DruidPotionEntity.java [hunk 1]
```diff
@@ -118,14 +118,22 @@ public void applySplash(List<MobEffectInstance> effects, @Nullable Entity entity
...
+        final Entity owner = super.getOwner();
+        if (owner instanceof AbstractEntityCitizen)
+        {
+            return (AbstractEntityCitizen)owner;
+        }
+        return null;
     }
-
-
+
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/entity/DruidPotionEntity.java:136]: This is possible via command lines.

**Review by marchermans — DISMISSED** [review]

File: src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java [hunk 2]
```diff
@@ -77,6 +79,32 @@
      */
     IBuilding getBuilding(BlockPos pos);
 
+    /**
+     * Get the immutable list of leisure sites.
```
Comment (someaddons) [inline src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java]: * get the positions of leisure sites, return type should be just List

File: src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java [hunk 3]
```diff
@@ -77,6 +79,32 @@
...
+     * @return the list.
+     */
+    ImmutableList<BlockPos> getLeisureSites();
+
+    /**
+     * Get the first building instance of the type.
```
Comment (someaddons) [inline src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java]: Get first building matching the conditions?
Method name probably should be more like `getFirstBuildingMatching` since the predicate is not exclusively to type

File: src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java [hunk 4]
```diff
@@ -148,6 +156,22 @@ public void read(@NotNull final CompoundTag compound)
                 addField(BlockPosUtil.read(fieldTagList.getCompound(i), TAG_POS));
             }
         }
+
+        if (compound.getAllKeys().contains(TAG_LEISURE))
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java]: a simple compound.contains is enough

File: src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java [hunk 5]
```diff
@@ -148,6 +156,22 @@ public void read(@NotNull final CompoundTag compound)
...
+        {
+            // Fields before Buildings, because the Farmer needs them.
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java]: copy pasted comment

File: src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java [hunk 6]
```diff
@@ -308,6 +334,93 @@ public IBuilding getBuilding(final BlockPos buildingId)
...
     }
 
+    @Override
+    public ImmutableList<BlockPos> getLeisureSites()
+    {
+        return leisureSites;
+    }
+
+    @Override
+    public BlockPos getRandomLeisureSite(final Random random)
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java]: Is there a reason for passing a random param along here? if not we could probably just use the static instance of the colony itself instead of having to find one to pass along each time we want to retrive a position

File: src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java [hunk 7]
```diff
@@ -308,6 +334,93 @@ public IBuilding getBuilding(final BlockPos buildingId)
...
+    public BlockPos getFirstBuildingOfType(final Predicate<IBuilding> predicate)
+    {
+        for (final IBuilding building : buildings.values())
+        {
+            if (predicate.test(building))
+            {
+                return building.getPosition();
+            }
+        }
+        return null;
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java:400]: might be better to return Blockpos.Zero instead

File: src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java [hunk 8]
```diff
@@ -308,6 +334,93 @@ public IBuilding getBuilding(final BlockPos buildingId)
...
+    }
+
+    @Override
+    public void addLeisureSite(final BlockPos pos)
+    {
+        final List<BlockPos> tempList = new ArrayList<>(leisureSites);
+        tempList.add(pos);
+        this.leisureSites = ImmutableList.copyOf(tempList);
+        colony.markDirty();
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java]: wouldnt it make more sense to do markBuildingsDirty()?

File: src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java [hunk 9]
```diff
@@ -308,6 +334,93 @@ public IBuilding getBuilding(final BlockPos buildingId)
...
+
+    @Override
+    public void removeLeisureSite(final BlockPos pos)
+    {
+        if (leisureSites.contains(pos))
+        {
+            final List<BlockPos> tempList = new ArrayList<>(leisureSites);
+            tempList.remove(pos);
+            this.leisureSites = ImmutableList.copyOf(tempList);
+            colony.markDirty();
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java]: same here with markBuildingsDirty()

File: src/main/java/com/minecolonies/coremod/entity/ai/minimal/EntityAICitizenWander.java [hunk 10]
```diff
@@ -1,83 +1,243 @@
...
+     */
+    private final TickRateStateMachine<WanderState> stateMachine;
+
+    /**
+     * Position to path to.
+     */
+    private BlockPos walkTo;
+
+    /**
+     * Position to path to.
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/entity/ai/minimal/EntityAICitizenWander.java]: copy pasted doc

File: src/main/java/com/minecolonies/coremod/entity/ai/minimal/EntityAICitizenWander.java [hunk 11]
```diff
@@ -1,83 +1,243 @@
...
     @Override
-    public boolean canContinueToUse()
+    public boolean canUse()
     {
-        return !citizen.getNavigation().isDone();
+        if (citizen.getDesiredActivity() == DesiredActivity.SLEEP)
+        {
+            return false;
+        }
+        stateMachine.tick();
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/entity/ai/minimal/EntityAICitizenWander.java]: sure you want to tick the AI before this returns true?

File: src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java [hunk 12]
```diff
@@ -308,6 +334,93 @@ public IBuilding getBuilding(final BlockPos buildingId)
...
+                return building.getPosition();
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public void addLeisureSite(final BlockPos pos)
+    {
+        final List<BlockPos> tempList = new ArrayList<>(leisureSites);
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java:406]: maybe use a Set on adding to prevent double values getting in?

File: src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java (hunk 3 above)
Comment (Raycoms) [reply src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java]: yeah, had changed the params after making the jdoc

File: src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java (hunk 7 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java:400]: BlockPos.Zero is nowadays a valid location, so I prefer returning null to make sure no one tries to use this position

Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java:400]: nah Blockpos.Zero is not valid, (0,0,0) is. You use BlockPos.Zero as invalid reference

File: src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java (hunk 12 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java:406]: will add a contains check, want to avoid using a set, because I can't do as easy of a random lookup with them

File: src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java (hunk 6 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java]: Wanted to avoid going through the colony -> world to get the random and a potential nullable

File: src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java (hunk 7 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java:400]: Yeah, but if I do == comparisons, then I can also do a == null comparison

Comment (uecasm) [reply src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java:400]: That's fragile as reference equality (or lack thereof) won't survive serialization (file or network).  I'm with Ray on that one.

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java:400]: This way, no one that calls this will think it's a valid pos.

File: src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java (hunk 6 above)
Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java]: Maybe we should just define one static random instance in a constant class somewhere and use that

File: src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java (hunk 12 above)
Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java:406]: oh I just meant new Set(existing).add(pos) and then make it back to a list, but contains works too

**Review by someaddons — DISMISSED** [review]

File: src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java [hunk 13]
```diff
@@ -77,6 +78,32 @@
      */
     IBuilding getBuilding(BlockPos pos);
 
+    /**
+     * Get the leisure site positions.
+     * @return the list.
+     */
+    List<BlockPos> getLeisureSites();
+
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java:86]: Split apart into a leisure manager.
Leisure sites are not directly related to buildings so they should be split apart.

File: src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java [hunk 14]
```diff
@@ -77,6 +78,32 @@
...
+     * @param pos the position of it.
+     */
+    void addLeisureSite(BlockPos pos);
+
+    /**
+     * Remove a leisure site.
+     * @param pos the position of it.
+     */
+    void removeLeisureSite(BlockPos pos);
+
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java:106]: See note above.

File: src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java [hunk 15]
```diff
@@ -302,4 +329,10 @@
      * @param level    The new level.
      */
     void onBuildingUpgradeComplete(@Nullable IBuilding building, int level);
+
+    /**
+     * Get a random leisure site to go to.
+     * @return the position of it.
+     */
+    BlockPos getRandomLeisureSite();
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java:337]: See note above.

File: src/main/java/com/minecolonies/coremod/blocks/BlockDecorationController.java [hunk 16]
```diff
@@ -114,6 +125,30 @@ public InteractionResult use(
...
+            return;
+        }
+
+        final BlockEntity tileEntity = worldIn.getBlockEntity(pos);
+        if (tileEntity instanceof TileEntityDecorationController && ((TileEntityDecorationController) tileEntity).getPositionedTags().getOrDefault(BlockPos.ZERO, new ArrayList<>()).contains(LEISURE))
+        {
+            @Nullable final IColony colony = IColonyManager.getInstance().getColonyByPosFromWorld(worldIn, pos);
+            if (colony != null)
+            {
+                colony.getBuildingManager().addLeisureSite(pos);
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/blocks/BlockDecorationController.java:147]: See the note on `IBuildingManager` this is not directly related to a building, so move it to a separate manager to clean up the API.

File: src/main/java/com/minecolonies/coremod/client/render/RenderBipedCitizen.java [hunk 17]
```diff
@@ -123,7 +123,7 @@ protected void renderNameTag(
             double distance = this.entityRenderDispatcher.distanceToSqr(entityIn.getX(), entityIn.getY(), entityIn.getZ());
             if (distance <= 4096.0D)
             {
-                double yOffset = model.young ? -0.8 : 0;
+                double yOffset = model.young ? -0.3 : 0;
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/client/render/RenderBipedCitizen.java:126]: Undocumented, unrelated change. Please document.

File: src/main/java/com/minecolonies/coremod/colony/buildings/modules/GraveyardManagementModule.java [hunk 18]
```diff
@@ -144,6 +144,23 @@ public GraveData getLastGraveData()
...
+        for (final String citizen : citizens)
+        {
+            if (restingCitizen.contains(citizen))
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/colony/buildings/modules/GraveyardManagementModule.java:163]: Why was this added?

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingFlorist.java [hunk 19]
```diff
@@ -149,15 +150,14 @@ public void removePlantableGround(final BlockPos pos)
...
+          .filter(stack -> !getModuleMatching(ItemListModule.class, m -> m.getId().equals(BUILDING_FLOWER_LIST)).isItemInList(stack)).toList();
+
         if (stacks.isEmpty())
         {
             return null;
         }
 
-        Collections.shuffle(stacks);
-        return stacks.get(0).getItemStack();
+        return stacks.get(MathUtils.RANDOM.nextInt(stacks.size())).getItemStack();
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingFlorist.java:160]: You can pull a random in one call, instead of copying it into a list...

Adding:
`.sorted((o1, o2) -> ThreadLocalRandom.current().nextInt(-1, 2))
            .findAny()
`
in the stream chain will return an entry without putting it in the list first.

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingFlorist.java [hunk 20]
```diff
@@ -149,15 +150,14 @@ public void removePlantableGround(final BlockPos pos)
     public ItemStack getFlowerToGrow()
     {
         final List<ItemStorage> stacks = getPlantablesForBuildingLevel(getBuildingLevel()).stream()
-                                           .filter(stack -> !getModuleMatching(ItemListModule.class, m -> m.getId().equals(BUILDING_FLOWER_LIST)).isItemInList(stack))
-                                           .collect(Collectors.toList());
+          .filter(stack -> !getModuleMatching(ItemListModule.class, m -> m.getId().equals(BUILDING_FLOWER_LIST)).isItemInList(stack)).toList();
+
         if (stacks.isEmpty())
         {
             return null;
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingFlorist.java:157]: No, turn this into an Empty stack, or make the method return an optional!

**Review by marchermans — CHANGES_REQUESTED** [review]
Generally an okey feature, implementation is food.

However for future reference: It is a lot easier to review this if you did not cram all those bug fixes in there......... So please refrain from doing that the next time and just take that additional minute to create a separate branch and PR..... You can even do it straight from IDEA, which means it takes even less time.

I would like to see the leisure api moved to a separate manager though.

File: src/main/java/com/minecolonies/coremod/colony/buildings/modules/GraveyardManagementModule.java (hunk 18 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/buildings/modules/GraveyardManagementModule.java:163]: Was going to add the additional things to the PR doc =P

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingFlorist.java (hunk 19 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingFlorist.java:160]: That is code that is not obvious to the reader and has even a chance to perform worse than this. I did research quite a lot about random elements from a stream and it's not super reliable.

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingFlorist.java (hunk 20 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingFlorist.java:157]: I'm not changing the logic here, just a hyper expensive shuffle call.

File: src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java (hunk 13 above)
Comment (Raycoms) [reply src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java:86]: I don't fully agree on this. The leisure sites function pretty exactly like buildings because.
a) Some buildings are leisure sites
b) Leisure sites use decoration controllers which are very similar to buildings.
c) Due to a) we query several building related things with this

Comment (marchermans) [reply src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java:86]: The key words here being **very similar but not equal **

Leisure sites are not buildings. Period.
Move it to a separate API. It would have taken you just as long to do that as it took you to write that message...

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingFlorist.java (hunk 19 above)
Comment (marchermans) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingFlorist.java:160]: True, i wish there was a better approach to this

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingFlorist.java (hunk 20 above)
Comment (marchermans) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingFlorist.java:157]: Make a Todo out of it then
Null itemstacks are forbidden you know that

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingFlorist.java:157]: If the method is nullable then it's fine, I don't see the necessity to handle around optionals and having to unpack them constantly. the null here indicates that sth is wrong anyway,

File: src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java (hunk 13 above)
Comment (Raycoms) [reply src/api/java/com/minecolonies/api/colony/managers/interfaces/IBuildingManager.java:86]: Then I rename the manager to "RegisteredStructureManager" because it also takes care of fields forever. + Leisure is a mix of buildings and deco. It makes sense to have them together.
